### PR TITLE
Sound: Handle control Queue

### PR DESCRIPTION
**Pull request by dorindabassey**

### CODE REVIEW

File: crates/sound/src/vhu_sound.rs [hunk 1]
```diff
@@ -2,22 +2,70 @@
...
+    #[error("Invalid descriptor count {0}")]
+    UnexpectedDescriptorCount(usize),
+    #[error("Invalid descriptor size, expected: {0}, found: {1}")]
+    UnexpectedDescriptorSize(usize, usize),
+    #[error("Invalid descriptor size, expected at least: {0}, found: {1}")]
+    UnexpectedMinimumDescriptorSize(usize, usize),
+    #[error("Received unexpected readable descriptor at index {0}")]
+    UnexpectedReadableDescriptor(usize),
+    #[error("Received unexpected write only descriptor at index {0}")]
+    UnexpectedWriteOnlyDescriptor(usize),
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: For now I think we can add the new errors directly in `lib.rs`

File: crates/sound/src/vhu_sound.rs [hunk 2]
```diff
@@ -68,6 +118,23 @@ impl VhostUserSoundThread {
...
 
     fn process_control(&self, _vring: &VringRwLock) -> IoResult<bool> {
+
+        if self.event_idx {
+            // vm-virtio's Queue implementation only checks avail_index
+            // once, so to properly support EVENT_IDX we need to keep
+            // calling process_request_queue() until it stops finding
+            // new requests on the queue.
+            loop {
+                _vring.disable_notification().unwrap();
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: the `_` in `_vring` is added when the parameter is not used, now you can remove it

File: crates/sound/src/vhu_sound.rs [hunk 3]
```diff
@@ -68,6 +118,23 @@ impl VhostUserSoundThread {
     }
 
     fn process_control(&self, _vring: &VringRwLock) -> IoResult<bool> {
+
+        if self.event_idx {
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: Maybe we can do this check in `handle_event()`

File: crates/sound/src/vhu_sound.rs [hunk 4]
```diff
@@ -82,6 +149,57 @@ impl VhostUserSoundThread {
     fn process_rx(&self, _vring: &VringRwLock) -> IoResult<bool> {
         Ok(false)
     }
+
+
+    pub fn process_requests(
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: IIUC in some way you would like to reuse this function for all the queues, but I think it is difficult, for now I would move all the code in `process_control`

File: crates/sound/src/vhu_sound.rs [hunk 5]
```diff
@@ -113,14 +233,15 @@ impl VhostUserSoundBackend {
...
 
+        debug!("everything is ok");
         Ok(Self {
             threads,
             virtio_cfg: VirtioSoundConfig {
                 jacks: 0.into(),
                 streams: 1.into(),
                 chmpas: 0.into(),
             },
-            exit_event: EventFd::new(EFD_NONBLOCK).map_err(Error::EventFdCreate)?,
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: Take a look at `lib.rs` on how to define `EventFdCreate`.
As I already suggested, I think we can avoid for now to add the new error enum here, but we can reuse the one in lib.rs

File: crates/sound/src/vhu_sound.rs [hunk 6]
```diff
@@ -82,6 +149,57 @@ impl VhostUserSoundThread {
...
+        _vring: &VringRwLock,
+    ) -> Result<bool> {
+        if requests.is_empty() {
+            debug!("yes, it's empty");
+            return Ok(true);
+        }
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            debug!("Sound request with n descriptors: {}", descriptors.len());
+            if descriptors.len() != 2 {
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: We can add the reference to the virtio spec as a comment to simplify the review

File: crates/sound/src/vhu_sound.rs (hunk 5 above)
Comment (dorindabassey) [reply crates/sound/src/vhu_sound.rs]: Ah yes yes, Ok we can reuse the one in lib.rs

File: crates/sound/src/vhu_sound.rs (hunk 3 above)
Comment (dorindabassey) [reply crates/sound/src/vhu_sound.rs]: agreed

File: crates/sound/src/vhu_sound.rs [hunk 7]
```diff
@@ -50,25 +53,181 @@ impl VhostUserSoundThread {
...
+        debug!("Requests to process: {}", requests.len());
+        if requests.is_empty() {
+            debug!("yes, it's empty");
+            return Ok(true);
+        }
+        //iterate over each sound request
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            debug!("Sound request with n descriptors: {}", descriptors.len());
+            if descriptors.len() != 3 {
```
Comment (MatiasVara) [inline crates/sound/src/vhu_sound.rs]: Since the number of descriptors depends on the request, I think the number of descriptors should be checked when evaluating the different control queue messages. `VIRTIO_SND_R_PCM_SET_PARAMS`, for example, has two descriptors.

Comment (dorindabassey) [reply crates/sound/src/vhu_sound.rs]: yes, make sense

File: crates/sound/src/vhu_sound.rs [hunk 8]
```diff
@@ -25,7 +28,7 @@ struct VhostUserSoundThread {
 }
 
 impl VhostUserSoundThread {
-    pub fn new(mut queue_indexes: Vec<u16>) -> Result<Self> {
+    pub fn new(mut queue_indexes: Vec<u16>) -> Result<Self, io::Error> {
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: Using `crate::Result`, we don't need to add this everywhere

File: crates/sound/src/vhu_sound.rs [hunk 9]
```diff
@@ -2,17 +2,20 @@
 
 use crate::audio_backends::{alloc_audio_backend, AudioBackend};
 use crate::virtio_sound::*;
-use crate::{Error, Result, SoundConfig};
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs:5]: I think we should avoid to remove `Result` here

File: crates/sound/src/vhu_sound.rs [hunk 10]
```diff
@@ -2,17 +2,20 @@
...
-
+use crate::{Error, SoundConfig};
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: in `rust-vmm` they usually organize the `use` section in this way:

```
use std::...
use std::mem::..

use something::...

use crate::...
```

Leaving a blank line to separate each section.

I think the section was in a wrong order before this commit (my bad), so feel free to don't fix here, but please leave the space between them.

File: crates/sound/src/vhu_sound.rs [hunk 11]
```diff
@@ -50,24 +53,244 @@ impl VhostUserSoundThread {
     }
 
     fn update_memory(&mut self, mem: GuestMemoryAtomic<GuestMemoryMmap>) -> IoResult<()> {
+        debug!("update memory");
         self.mem = Some(mem);
         Ok(())
     }
 
-    fn handle_event(&self, device_event: u16, vrings: &[VringRwLock]) -> IoResult<bool> {
+    fn handle_event(&self, device_event: u16, vrings: &[VringRwLock]) -> result::Result<bool, io::Error> {
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: Why this change is needed?

File: crates/sound/src/vhu_sound.rs [hunk 12]
```diff
@@ -50,24 +53,244 @@ impl VhostUserSoundThread {
...
+                        .write_obj(response, desc_response.addr())
+                        .map_err(|_| Error::DescriptorWriteFailed)?;
+                    let len = desc_response.len() as u32;
+
+                    if vring
+                        .add_used(desc_chain.head_index(), len)
+                        .is_err()
+                        {
+                            error!("Couldn't return used descriptors to the ring");
+                        }
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: This seems a common pattern, can we move it after the `match` ?

File: crates/sound/src/virtio_sound.rs [hunk 13]
```diff
@@ -92,6 +92,12 @@ pub const VIRTIO_SND_PCM_FMT_DSD_U16: u8 = 22;
 pub const VIRTIO_SND_PCM_FMT_DSD_U32: u8 = 23;
 pub const VIRTIO_SND_PCM_FMT_IEC958_SUBFRAME: u8 = 24;
 
+
+pub const SUPPORTED_FORMATS: u64 = 1 << VIRTIO_SND_PCM_FMT_U8
```
Comment (stefano-garzarella) [inline crates/sound/src/virtio_sound.rs]: I would move this out of `virtio_sound.rs` where we should put only the stuff defined into the virtio-spec

File: crates/sound/src/virtio_sound.rs [hunk 14]
```diff
@@ -109,6 +115,14 @@ pub const VIRTIO_SND_PCM_RATE_176400: u8 = 11;
 pub const VIRTIO_SND_PCM_RATE_192000: u8 = 12;
 pub const VIRTIO_SND_PCM_RATE_384000: u8 = 13;
 
+pub const SUPPORTED_RATES: u64 = 1 << VIRTIO_SND_PCM_RATE_8000
```
Comment (stefano-garzarella) [inline crates/sound/src/virtio_sound.rs]: Ditto.

**Review by stefano-garzarella — CHANGES_REQUESTED** [review]
@dorindabassey thanks for the PR!

I think it is in a good state, I just pointed out some minor things to fix, then we can merge it!

File: crates/sound/src/vhu_sound.rs (hunk 10 above)
Comment (dorindabassey) [reply crates/sound/src/vhu_sound.rs]: ok, noted. Thanks :)

File: crates/sound/src/virtio_sound.rs (hunk 13 above)
Comment (dorindabassey) [reply crates/sound/src/virtio_sound.rs]: do you have any preference on where to put it? I think lib.rs looks like a good place

Comment (stefano-garzarella) [reply crates/sound/src/virtio_sound.rs]: In `vhu_sound.rs` should be fine for now

File: crates/sound/src/vhu_sound.rs [hunk 15]
```diff
@@ -1,23 +1,40 @@
...
 use vmm_sys_util::{
     epoll::EventSet,
     eventfd::{EventFd, EFD_NONBLOCK},
 };
 
+use crate::audio_backends::{alloc_audio_backend, AudioBackend};
+use crate::virtio_sound::*;
+use crate::{Error, Result, SoundConfig};
+
+pub const SUPPORTED_FORMATS: u64 = 1 << VIRTIO_SND_PCM_FMT_U8
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs:25]: It seems `SUPPORTED_FORMATS` and `SUPPORTED_RATES` are used only in this module, so we can remove `pub`
